---
title: "Prime vs TIRE-seq comparison"
description: "Compare quality control metrics"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6,
                      out.width="150%",
                      fig.path='../figures/compare_prime_tire_metrics/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())

library(tidyverse)
library(platetools)
library(patchwork)
library(edgeR)
library(scuttle)
library(scater)
library(here)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Prime vs TIRE-seq compare metrics

## Recap

I sequenced these samples with Prime-seq on extracted RNAs already. Results were not that great. Here reprocess the same samples with TIRE-seq for a head to head comparison.

## Samples

* Yasmin Nouri plates.
* All wells have 10,000 cells in 50ul total volume.
* Comprised of 25ul cells in media and 25ul 2x buffer TCL
* The cell concentration is therefore 200 cells/uL.
* So 20uL input will be 4000 cells.

## Lab notes

The processing went as planned. Full writeup available at [ELN link](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/NDkxLjQwMDAwMDAwMDAwMDAzfDY4MTk3LzM3OC9UcmVlTm9kZS8zMzc4NjI0NXwxMjQ3LjM5OTk5OTk5OTk5OTk=)

## Downsampling notes

This was doone in the notebook scripts */S000553/downsample_humanTcell.sh*

Prime-seq = MB2_PRM0003_YN sequenced on 2 runs:  
* S000514 40M
* S000493 149.8M
* Combined = 189.8M / 192 samples = 988k reads per well
* The object I import was generated in G000233_DRAC_notebooks/MB2_Tcell/1A_generate_SCE.rmd

TIRE-Seq

* S000553 160.0M / 96 samples = 1.66M reads per well
* Therefore need to downsample this to 988k * 96 = **94,848,000 reads** for the library

## Read SCEs

This is generated in 1A notebook generate SCE.

```{r}
tire <- readRDS(here::here(
  "data/TIRE_Tcell/SCEs/Downsample/tire_tcell_ds_basic.sce.rds"
))

prime <- readRDS(here::here(
  "data/TIRE_Tcell/SCEs/Downsample/prime_tcell_ds_basic.sce.rds"
))
```

Extract the sample metadata and unify before merging into a tibble.  
Keep only the untransduced normal T cells in the Prime-seq experiment.

```{r}
tire_tb <- as_tibble(colData(tire))
tire_tb$Protocol <- "TIRE"

prime_tb <- as_tibble(colData(prime))
prime_tb$Protocol <- "Prime"

keep_cols <- intersect(
  colnames(tire_tb),
  colnames(prime_tb)
)

tb <- rbind(
  tire_tb[,keep_cols],
  prime_tb[,keep_cols]
)
```

# Quality control metrics {.tabset}

All metrics are better for TIRE-seq

## Reads and UMIs

Better for TIRE-seq

```{r saturation_log}
plt1 <- ggplot(data=tb, aes(y=sum+1, x=Reads+1, colour=Protocol)) +
  geom_point() +
  scale_colour_brewer(type="qualitative", palette = "Dark2") +
  xlab("Reads") + ylab("Library size (UMIs)") +
  scale_y_continuous(trans='log10', limits = c(1, 1e7)) +
  scale_x_continuous(trans='log10', limits = c(1, 1e7)) +
  annotation_logticks(base = 10, sides = "bl") +
  coord_fixed(ratio = 1)

plt1
```

## UMIs per timepoint

This show the activation of transcription at day 1 and 2

```{r umi_time}
tb <- tb %>% 
  mutate(Day = as.numeric(str_split(pattern = "_", Timepoint, simplify = TRUE)[,2]))

plt5 <- ggplot(data = tb, aes(y = sum + 1, x = Day, colour = Protocol)) +
  geom_point() +
  geom_smooth(se = TRUE, method = "loess", span = 0.6) + # Adjust `span` for better fit
  scale_colour_brewer(type = "qualitative", palette = "Dark2") +
  xlab("Day") + 
  ylab("Library size (UMIs)") +
  scale_y_continuous(trans = 'log10', limits = c(1e3, 1e7)) +
  annotation_logticks(base = 10, sides = "l")

plt5
```

## Mapping stats

Better for TIRE-seq.

```{r mapping}
mapping <- tb %>% 
  select(Intergenic:Reads, Protocol,Well) %>% 
  pivot_longer(cols = c(Intergenic, Exon, Ambiguity, Unmapped),
               names_to = "Feature",
               values_to = "Count") %>%
  mutate(Percentage = Count / Reads * 100)

plt2 <- ggplot(mapping,
             aes(x = Feature, y= Percentage, colour = Protocol)) + 
  geom_boxplot() +
  ylab("Percent") + ylim(0,100) +
  xlab("") +
  scale_colour_brewer(palette = "Dark2")

plt2
```

## Mitochondrial gene percent

Very low which is good.

```{r mito, fig.asp=1}
plt3 <- ggplot(tb,
             aes(x = Protocol, y= subsets_Mito_percent, colour = Protocol)) + 
  geom_boxplot() + geom_jitter() +
  ylab("Mitochondrial %") + 
  xlab("") +
  scale_colour_brewer(palette = "Dark2")

plt3
```

## Genes detected

Very low which is good.

```{r genes, fig.asp=0.8}
plt4 <- ggplot(tb,
             aes(x = Protocol, y= detected, colour = Protocol)) + 
  geom_boxplot() + geom_jitter() +
  ylab("Genes detected") + 
  xlab("") +
  scale_colour_brewer(palette = "Dark2")

plt4
```


# Gene correlation

```{r rowSums}
rowData(tire)$sum <- rowSums2(counts(tire))
rowData(tire)$protocol <- "TIRE"

rowData(prime)$sum <- rowSums2(counts(prime))
rowData(prime)$protocol <- "Prime"

gene_tb <- as_tibble(rbind(
  rowData(tire),
  rowData(prime)
))

# Add 1 before logging
gene_tb$sum <- gene_tb$sum + 1
```

Convert to wide tibble

```{r}
gene_tb_wide <- gene_tb %>%
  select(ID,Symbol,sum,protocol) %>% 
  pivot_wider(names_from = protocol, values_from = sum)

gene_tb_wide

gene_tb_wide_clean <- na.omit(gene_tb_wide)
correlation <- cor(log10(gene_tb_wide_clean$Prime), log10(gene_tb_wide_clean$TIRE), method = "pearson")
```

## Plot the result of gene gene correlation

The correlation is `r correlation`

```{r gene_gene_sum, fig.asp=1}
plt4 <- ggplot(gene_tb_wide_clean, 
             aes(x = Prime, y= TIRE)) + 
  geom_point(alpha = 0.33, size=1) + 
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1))) +
  xlab("Prime-seq") + 
  ylab("TIRE-seq") +
  ggtitle("Counts by gene") +
  geom_smooth(method = "lm", formula = y ~ x, color = "red", 
              se = TRUE, level = 0.95) +
  scale_y_continuous(trans='log10', limits = c(1, 1e6)) +
  scale_x_continuous(trans='log10', limits = c(1, 1e6)) +
  annotation_logticks(base = 10, sides = "bl") +
  theme_Publication(base_size = 18)

plt4
```

# Differential expression testing

Check what genes are different between Prime and TIRE-seq particularly some outliers detected in Prime-seq.

First need to combine the objects.

```{r}
keep_genes <- intersect(
  row.names(tire),
  rownames(prime)
)

counts <- cbind(
  counts(tire[keep_genes]),
  counts(prime[keep_genes])
)

dge <- DGEList(counts = counts, 
               group = tb$Protocol)
```

Perform preprocessing steps

```{r}
keep <- filterByExpr(dge, group=dge$samples$Timepoint, min.count=1)
dge <- dge[keep,]
dge <- calcNormFactors(dge)
design <- model.matrix(~group, data=dge$samples)
dge <- estimateDisp(dge, design)
fit <- glmQLFit(dge, design)
```

Perform the DE test

```{r}
qlf <- glmQLFTest(fit, coef = 2)
results <- as_tibble(topTags(qlf, n = Inf)$table)
results$Symbol <- row.names(topTags(qlf, n = Inf)$table)
sig_genes <- results %>%
  filter(FDR < 0.05, abs(logFC) > 1)

sig_genes
```

Create the MA plot

```{r ma_plot}
ggplot(results, aes(x = logCPM, y = logFC)) +
  geom_point(aes(color = FDR < 0.05), alpha = 0.3, size=0.75) +
  geom_hline(yintercept = 0, color = "red", linetype = "dashed") +
  scale_color_manual(values = c("grey", "darkblue")) +
  labs(
    x = "Log2 CPM",
    y = "Log2 FC",
    color = "Significant"
  ) +
  theme_Publication()
```

# Plot the technical variation

I generated this with ChaptGPT io

I could not use edgeR BCV function bacause Prime-seq had too few counts to pass the filter by expression step.

```{r}
# Start with unfiltered dgelist object
dge <- DGEList(counts = counts, 
               group = tb$Protocol)
# Calculate normalized counts (log-transformed counts per million)
logCPM <- edgeR::cpm(dge, log=TRUE, normalized.lib.sizes=TRUE)
# Convert to a data frame
logCPM_df <- as.data.frame(logCPM)
logCPM_df$Gene <- rownames(logCPM_df)

# Reshape the data to long format
library(reshape2)
logCPM_long <- melt(logCPM_df, id.vars = "Gene", variable.name = "Sample", value.name = "Expression")

# Add group information to the data
logCPM_long$Group <- dge$samples$group[match(logCPM_long$Sample, rownames(dge$samples))]

# Calculate mean expression per gene within each group
gene_stats <- logCPM_long %>%
  group_by(Gene, Group) %>%
  summarize(
    Variance = var(Expression, na.rm = TRUE),
    MeanExpression = mean(Expression, na.rm = TRUE)
  )
```

## Plot variance vs. mean expression

The shape here is typical of **unfiltered** counts which is what I supplied to the plot.  
If I filter counts I lose too many genes from Prime-seq.

```{r tech_var}
ggplot(gene_stats, aes(x=MeanExpression, y=Variance, color=Group)) +
  geom_point(alpha=0.5) +
  theme_Publication() +
  scale_colour_brewer(palette = "Dark2") +
  xlab("Mean Genewise log-CPM") + ylab("Variance") + labs(color="Protocol")
```

# Conclusion

TIRE-seq better in all metrics

## Session info
```{r}
sessionInfo()
```