---
title: "TIRE cell input mapping stats"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/cellInput_mapping/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())

library(SingleCellExperiment)
library(tidyverse)
library(scuttle)
library(scater)
library(platetools)
library(patchwork)
library(here)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# TIRE cell input mapping stats

## Aim

Compare the mapping to certain features of the genome (exon, intron etc) between Prime-seq and TIRE-seq.

## Samples
 
A serial dilution will be prepared on the day of the experiment.

* Human HEK293T
* Mouse NIH3T3
* No template control

[ELN entry](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/NDMyLjkwMDAwMDAwMDAwMDAzfDY4MTk3LzMzMy9UcmVlTm9kZS8yMjA0NzM2OTgzfDEwOTguODk5OTk5OTk5OTk5OQ==)

Standard TSO where UMI on TSO followed by spacer. Well barcode in line with read 1

## Data to read in

### Prime-seq

This is an experiment where I compared different inputs.

* Purified RNA from silica column
* 100 Intact FACS sorted cells
* Crude lysate without any purification, conc 1000 cells per uL

```{r}
prime <- readRDS(here::here(
  "data/NN191/mappingByAnnotation.rds"
)) %>% 
  select(type,Count,percent,Sample_type)

colnames(prime) <- c("RG", "type", "N", "Percent", "Input")
prime$Protocol <- "Prime-seq"
prime$Input <- recode(prime$Input, 
                      "Lysate" = "Prime Lysate",
                      "Cells" = "Prime Intact Cells",
                      "RNA" = "Prime Purified RNA")
prime
```

### TIRE-seq

I sequenced these libraries with the configuration

* Read 1 70bp
* Read 2 52bp
* Index read 1 and 2 8bp each

```{r}
meta_data <- read_csv(here::here(
  "metadata/S000442/DCs_Bcor_metadata_v1.1_HS.csv"))

mapping <- read_tsv(here::here(
  "data/S000442/stats/TURBO_DC_Bcor_HsMm.readspercell.txt"
))

tire <- inner_join(mapping, meta_data,
                    by=c("RG" = "Well_BC")) %>% 
  filter(Researcher == "DB") %>% 
  filter(Cell_type != "NTC") %>% 
  select(RG:type, Cell_type) %>% 
  group_by(RG) %>% 
  mutate(
    Reads = sum(N),
    Percent = (N / Reads) * 100
    )

tire$Protocol <- "TIRE-Seq"
tire$Input <- "TIRE Lysate"
colnames(tire) <- c("RG", "N", "type", "Cell_type", "Reads", "Percent", "Protocol", "Input")

tire
```

Combine the tibbles together

```{r}
tb <- rbind(prime, tire)
```

Visualise the outputs

```{r mapping}
ggplot(data=tb, aes(x=type, y=Percent, fill=Input)) +
  geom_boxplot(outlier.colour = "transparent") +
  scale_colour_brewer(type="qualitative", palette = "Dark2") +
  ylim(0,80) +
  xlab("Annotation") + ylab("Percent mapping")
```

# Conclusion

* Some elevated intergenic mapping compared to purified RNA.
* These are excluded when counting reads occurs.

### Session info
```{r}
sessionInfo()
```