---
title: "HEK293T RNA comparison"
description: "Protocol sensitivity comparison based on downsampling"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6,
                      fig.path='./figures/sensitivty_HEK/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())

library(tidyverse)
library(scuttle)
library(here)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication(base_size = 16))
```

# A protocol comparison of Prime-seq and TIRE-seq on HEK293Ts

We use human mapped data as that is what Janijic et al Prime-Seq is.  

```{r}
tb <- read_csv(here::here(
  "data/TIRE_technical/Matrices/HEK_downsampling_stats.csv"
))

tb
```

Plot less than 2.5M reads per sample. This is the range I used for 96w experiments.

```{r}
tb <- tb %>% 
  filter(reads <= 2.5e6)
```

# Visualise the results {.tabset}

Prime-Seq potentially deceptive as I got processed data mapped to a gencode reference. Clear that poor mapping rate in minibulk v2 means it loses to minibulk v1.

Minibulk v2 catches up as the saturation is less and we read through the primer dimers.

## Genes detected

TIRE-Seq detects more genes. This might be becuase of the 5' profiling. I may have to prove this.

```{r downsample_gene, fig.cap="Error bars are median absolute deviation"}
plt1 <- ggplot(data=tb, 
               aes(x=reads, y=med_gene, group=Protocol, color=Protocol)) +
  geom_line(size=1.25) + geom_point(size=2) +
  geom_errorbar(aes(ymin=med_gene-mad_gene, ymax=med_gene+mad_gene), width=.1, size=1.25) +
  scale_colour_brewer(type="Qualitative", palette = "Dark2") +
  xlab("Reads per sample") + ylab("Median genes detected")

plt1
```

## UMIs detected

Prime-Seq and TIRE-seq perform the same.

```{r downsample_umi, fig.cap="Error bars are median absolute deviation"}
plt2 <- ggplot(data=tb, 
               aes(x=reads, y=med_umi, group=Protocol, color=Protocol)) +
  geom_line(size=1.25) + geom_point(size=2) +
  geom_errorbar(aes(ymin=med_gene-mad_gene, ymax=med_gene+mad_gene), width=.1, size=1.25) +
  scale_colour_brewer(type="Qualitative", palette = "Dark2") +
  xlab("Reads per sample") + ylab("Median UMIs detected")

plt2
```

## Number of samples at depth

Prime-Seq and TIRE-seq stands out from the in house Prime-seq implementations.

```{r downsample_wells, fig.cap="Error bars are median absolute deviation"}
plt2 <- ggplot(data=tb, 
               aes(x=reads, y=cells, group=Protocol, color=Protocol)) +
    geom_line() + geom_point() +
  scale_colour_brewer(type="Qualitative", palette = "Dark2") +
  xlab("Reads per sample") + ylab("Number of samples")

plt2
```

# Investigate extra genes detected TIRE-seq

Reaad counts downsampled to 2M

#!!!HERE

```{r}
dat_prime <- here::here("data/GenomicsRnD/PrimeSeq/Bulk_opt_lysis_test_2_HEK.dgecounts.rds")
zUMI_prime <- readRDS(dat_prime)
zUMI_prime <- zUMI_prime$umicount$exon$downsampling$downsampled_2000000

dat_tire <- here::here("/stornext/Projects/GenomicsRnD/brown.d/S000514/zUMI_outputs/expression/TIRE_HEK_gencode.dgecounts.rds")
zUMI_tire <- readRDS(dat_tire)
zUMI_tire <- zUMI_tire$umicount$exon$downsampling$downsampled_2000000
```

Look at what extra genes are detected for TIRE-seq

```{r}
tire_prime <- intersect(
  row.names(zUMI_prime),
  row.names(zUMI_tire)
)

in_prime <- row.names(zUMI_prime)[!row.names(zUMI_prime) %in% tire_prime]
in_tire <- row.names(zUMI_tire)[!row.names(zUMI_tire) %in% tire_prime]
```

* The number of genes unique to Janjic Prime-seq is `r length(in_prime)`
* The number of genes unique to TIRE-seq is `r length(in_tire)`

### Session info
```{r}
sessionInfo()
```
