---
title: "UHRR protocol comparisons"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=1,
                      fig.path='./figures/downsampling/UHRR/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())

library(tidyverse)
library(scuttle)
library(here)
source(here::here("scripts/R/theme_publication.R"))
source(here::here("scripts/R/extract_num_features-subset.R"))
source(here::here("scripts/R/cal_seq_saturation.R"))
theme_set(theme_Publication())
```

# A protocol comparison of Minibulk and TurboCapture on UHRR

I reproduced the experiment from the Prime-Seq experiment where universal human reference RNA (UHRR) was mixed with ERCC spike-in RNAs with the protocols I implemented:

* Minibulk v2 (modified Prime-Seq)
* TIRE-Seq (Turbocapture Integrated RNA Expression)

## Read data

These are both downsampled to 10M reads per cell

```{r}
mb2 <- read_csv(here::here(
  "data/MB2_UHRR/Matrices/uhrr_coldata.csv"
))
mb2$Protocol <- "Primeseq"

tire <- read_csv("/stornext/Home/data/allstaff/b/brown.d/Projects/G000278_TurboCapSeq/data/TIRE_UHRR/Matrices/uhrr_coldata.csv")
tire$Protocol <- "TIREseq"

tb <- rbind(
  mb2,
  tire
)
```

## Library size genes etc {.tabset}

### Genes and UMIs

```{r reads_log, fig.asp=0.6}
plt8 <- ggplot(data=tb, aes(y=sum+1, x=detected, colour=Protocol)) +
  geom_point() +
  scale_colour_brewer(type="qualitative", palette = "Dark2") +
  xlab("Genes") + ylab("UMIs") +
  scale_y_continuous(trans='log10') +
  annotation_logticks(base = 10, sides = "l")

plt8
```

### Library size

```{r vio_library_user}
plt9 <- ggplot(tb,
             aes(x = Protocol, y= sum)) + 
  geom_violin() + geom_jitter(aes(colour=Protocol), width = 0.1) +
  ylab("Library Size (UMIs)") +
  xlab("Plate") +
  scale_y_continuous(trans='log10', limits = c(2e5,7e5)) +
  annotation_logticks(base = 10, sides = "l") +
  scale_colour_brewer(palette = "Dark2")

plt9 + theme(legend.position = "none")
```

## Library size ERCC

```{r vio_library_ERCC}
plt10 <- ggplot(tb,
             aes(x = Protocol, y= altexps_ERCC_sum)) + 
  geom_violin() + geom_jitter(aes(colour=Protocol), width = 0.1) +
  ylab("ERCC Library Size") + 
  xlab("") +
  scale_y_continuous(trans='log10', limits = c(1000,6000)) +
  annotation_logticks(base = 10, sides = "l") +
  scale_colour_brewer(palette = "Dark2")

plt10 + theme(legend.position = "none")
```

## Genes detected

```{r vio_genes}
plt10 <- ggplot(tb,
             aes(x = Protocol, y= detected)) + 
  geom_violin() + geom_jitter(aes(colour=Protocol), width = 0.1) +
  ylab("Genes") + ylim(10000,20000) +
  xlab("") +
  scale_colour_brewer(palette = "Dark2")

plt10 + theme(legend.position = "none")
```

### Total mapping rate

```{r mapping_vio}
plt9 <- ggplot(tb,
             aes(x = Protocol, y= Mapped)) + 
  geom_violin() + geom_jitter(aes(colour=Protocol), width = 0.1) +
  ylab("Mapping %") + ylim(0,100) +
  xlab("") +
  scale_colour_brewer(palette = "Dark2")

plt9 + theme(legend.position = "none")
```

### Exon mapping rate

```{r exon_mapping_vio}
plt9 <- ggplot(tb,
             aes(x = Protocol, y= Exon_Map)) + 
  geom_violin() + geom_jitter(aes(colour=Protocol), width = 0.1) +
  ylab("Exon mapping %") + ylim(0,100) +
  xlab("") +
  scale_colour_brewer(palette = "Dark2")

plt9+ theme(legend.position = "none")
```

## Sequencing saturation

Sequencing Saturation = 1 - (n_deduped_reads / n_reads)

From the plate metadata

```{r}
tb$Saturation <- 1 - (tb$sum / tb$Reads)

stats <- tb %>% 
  group_by(Protocol) %>% 
  summarise(
    mean_Saturation = mean(Saturation, na.rm=TRUE),
    median_Saturation = median(Saturation, na.rm=TRUE)
    )

stats
```

### Sequencing saturration

```{r saturation_vio}
plt9 <- ggplot(tb,
             aes(x = Protocol, y= Saturation)) + 
  geom_violin() + geom_jitter(aes(colour=Protocol), width = 0.1) +
  ylab("Sequencing saturation") + ylim(0,1) +
  xlab("") +
  scale_colour_brewer(palette = "Dark2")

plt9+ theme(legend.position = "none")
```

# Conclusions

Minibulk 2 has better metrics when sequenced deeply

### Session info
```{r}
sessionInfo()
```
