---
title: "Check sequecing saturation TIRE UHRR"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/checkSat/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())

library(tidyverse)
library(scuttle)
library(scran)
library(scater)
library(here)
library(patchwork)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# TurboCapture-Seq first test of 96 well barcode primers

## Recap

These are the first experiments where I use version 2 chemistry. Read 2 is now the barcode and UNI. Read 1 is the cDNA. This was done to avoid reading through the TSO on Ilumina read 1

Benchmark the version 2 protocol with a mixture of human and mouse cell lines Daniel Brown generated [https://au-mynotebook.labarchives.com/share/Piper_Research_Project/MjkuOTAwMDAwMDAwMDAwMDAyfDE3MTc2NS8yMy9UcmVlTm9kZS8yMzM4NDg4MTE1fDc1Ljg5OTk5OTk5OTk5OTk5](ELN link)

# Check stats data for reads and UMIs {.tabset}

This is the file emitted in the stats directory in zUMIs so is the most accurate.

```{r}
sce <- readRDS(here::here(
  "data/TIRE_UHRR/SCEs", "unhrr_basic.sce.rds"
))
stats_reads <- read_tsv(here::here(
  "data/GenomicsRnD/S000514/zUMI_outputs/stats/TIRE_UHRR_Hs.readspercell.txt"
))
colnames(stats_reads) <- c("Sample_BC", "Reads", "Feature")

stats_umis <- read_tsv(here::here(
  "data/GenomicsRnD/S000514/zUMI_outputs/stats/TIRE_UHRR_Hs.UMIcounts.txt"
))
colnames(stats_umis) <- c("UMIs", "Sample_BC", "Feature")

stats <- inner_join(stats_reads, stats_umis)
```

Check the saturation

```{r}
sat <- stats %>% 
  mutate(Saturation = 1 - (UMIs / Reads))
  
summary(sat$Saturation)
```

## Plot high expression

The limit of a 10nt UMI is `r 4^10` or ~ 1 million. 
The maximum here is 15,000 so it can't be UMI saturation

```{r high_expr}
plotHighestExprs(sce, n=10, as_percentage=FALSE) +
  theme_Publication()
```

Look at the individual top 4 genes

```{r}
plotExpression(sce, features = c("MT-RNR2", "MT-CO1", "MT-ND4", "RPS6"),exprs_values="counts") +
  theme_Publication()
```

### Session info
```{r}
sessionInfo()
```