---
title: "UHRR protocol correlation"
description: "What protocols are similar to what and check differences"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6,
                      fig.path='./figures/protocol_corr/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())

library(tidyverse)
library(scuttle)
library(here)
library(pheatmap)
library(edgeR)
library(viridis)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Correlate protocols of Prime-seq and TIRE-seq

In notebook 1C TIRE-seq detects more genes and similar UMIs as Prime-seq publication

Janjic, Aleksandar, Lucas E. Wange, Johannes W. Bagnoli, Johanna Geuder, Phong Nguyen, Daniel Richter, Beate Vieth et al. "Prime-seq, efficient and powerful bulk RNA sequencing." Genome biology 23, no. 1 (2022): 88.

## Aim

Investigate the similarities and differences between the different protocols.

## Read data {.tabset}

### Janjic et al Prime-seq

2 million read downsampling object is selected as this is the most Janjic et al., is sequenced. 

```{r}
dat_prime <- here::here("/stornext/Projects/GenomicsRnD/brown.d/PrimeSeq/prime-seq.dgecounts.rds")
zUMI_prime <- readRDS(dat_prime)
prime_counts <- zUMI_prime$umicount$exon$all

# Rename columns
colnames(prime_counts) <- paste("Janjic Prime", colnames(prime_counts), sep="_")
```

### In house Prime-seq

```{r}
dat_mb2 <- here::here("/stornext/Projects/GenomicsRnD/brown.d/S000514/zUMI_outputs/expression/MB2_UHRR_Gencode_Hs.dgecounts.rds")
zUMI_mb2 <- readRDS(dat_mb2)
mb2_counts <- zUMI_mb2$umicount$exon$all

# Keep the UHRR added colums only
mb2_counts <- mb2_counts[,colSums2(mb2_counts) > 5000]

# Rename columns
colnames(mb2_counts) <- paste("Modified Prime", colnames(mb2_counts), sep="_")
```

### Novel protocol TIRE-Seq

TIRE-seq which uses the Qiagen turbocapture plate

```{r turbo}
dat_turbo <- "/stornext/Projects/GenomicsRnD/brown.d/S000514/zUMI_outputs/expression/TIRE_UHRR_Gencode_Hs.dgecounts.rds"
zUMI_turbo <- readRDS(dat_turbo)
turbo_counts <- zUMI_turbo$umicount$exon$all

# Keep the UHRR added colums only
turbo_counts <- turbo_counts[,colSums2(turbo_counts) > 5000]

colnames(turbo_counts) <- paste("TIRE", colnames(turbo_counts), sep="_")
```

### Gold standard TruSeq RNA

This was requested by reviewer 2.

```{r truseq}
dat_truseq <- "/vast/projects/G000448_Protein_Design/BulkRNA_Manuscript/Public/UHRR/expression/TruSeq_concat_UHRR_public_Hs.dgecounts.rds"
zUMI_truseq <- readRDS(dat_truseq)
truseq_counts <- zUMI_truseq$readcount$exon$all

# Keep the UHRR added colums only
truseq_counts <- truseq_counts[,colSums2(truseq_counts) > 5000]

colnames(truseq_counts) <- paste("TruSeq", colnames(truseq_counts), sep="_")
```

# Correlate samples

Need to subset for genes that are in common with all protocols. Then I am able to combine them all together into a single count matrix.

# UP TO HERE

Consider removing the ERCCs as mine were really old.

```{r}
keep_genes <- intersect(
  row.names(prime_counts),
  row.names(mb2_counts)
)

keep_genes <- intersect(
  keep_genes,
  row.names(turbo_counts)
)

keep_genes <- intersect(
  keep_genes,
  row.names(truseq_counts)
)

diff_prime_turbo <- setdiff(
  row.names(prime_counts),
  row.names(turbo_counts)
)

diff_mb2turbo <- setdiff(
  row.names(mb2_counts),
  row.names(turbo_counts)
)

count_combo <- cbind(
  prime_counts[keep_genes,],
  mb2_counts[keep_genes,],
  turbo_counts[keep_genes,],
  truseq_counts[keep_genes,]
)
cpm_combo <- cpm(count_combo)

tb <- Matrix::as.matrix(cpm_combo) |>
  as.data.frame() |>
  tibble::as_tibble(rownames = "Gene_ID")

write_csv(tb, here::here(
  "data/TIRE_technical/Matrices/UHRR_sample_corr.csv"
))
```

Make an annotation bar for the heatmap.

```{r}
annotation_df <- as.data.frame(colnames(
  cpm_combo
))

annotation_df$Protocol <- str_split(string = annotation_df$`colnames(cpm_combo)`, pattern = "_", simplify = T)[,1]
annotation_df$Tag_count <- recode(annotation_df$Protocol,
                              "Janjic Prime" = "Three prime",
                              "Modified Prime" = "Three prime",
                              "TIRE" = "Five prime",
                              "TruSeq" = "Full length")

annotation_df <- annotation_df[,c(2:3)]
```

## Generate heatmap

Generate the correlation matrix then plot as a heatmap.

```{r heatmap}
cor_matrix <- cor(as.matrix(cpm_combo))
rownames(annotation_df) <- colnames(cor_matrix)
row.names(cor_matrix) <- str_split(colnames(cor_matrix), pattern = "_", simplify = T)[,1]

# Get Dark2 palette colors for Protocol
protocol_colors <- RColorBrewer::brewer.pal(4, "Dark2")
names(protocol_colors) <- c("Janjic Prime", "Modified Prime", "TIRE", "TruSeq")

# Distinct colors for Tag_count
tag_count_colors <- c("Five prime" = "#E69F00",    # Golden yellow
                     "Three prime" = "#56B4E9",    # Light blue
                     "Full length" = "#90EE90")     

# Combine into annotation colors list
annotation_colors <- list(
  Protocol = protocol_colors,
  Tag_count = tag_count_colors
)

pheatmap(cor_matrix,
         color = viridis(100),
         annotation_col = annotation_df,
         annotation_colors = annotation_colors,
         show_rownames = TRUE,
         show_colnames = FALSE,
         cluster_rows = FALSE,
         fontsize_number = 14,
         fontsize_row = 8)
```

Have a closer look specifically at the correlation between TruSeq and TIRE-seq

```{r}
summary(cor_matrix["TruSeq",c(17:24)])
```

Compare against the correlation between TruSeq and Janijic Prime

```{r}
summary(cor_matrix["TruSeq",c(1:8)])
```

* Mean of TruSeq and TIRE-Seq = 0.704
* Mean of TruSeq and Janjic Prime = 0.680


### Session info
```{r}
sessionInfo()
```
