---
title: "UHRR protocol correlation"
description: "What protocols are similar to what and check differences"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6,
                      fig.path='./figures/protocol_corr/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())

library(tidyverse)
library(scuttle)
library(here)
library(pheatmap)
library(viridis)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# Correlate protocols of Prime-seq and TIRE-seq

In notebook 1C TIRE-seq detects more genes and similar UMIs as Prime-seq publication

Janjic, Aleksandar, Lucas E. Wange, Johannes W. Bagnoli, Johanna Geuder, Phong Nguyen, Daniel Richter, Beate Vieth et al. "Prime-seq, efficient and powerful bulk RNA sequencing." Genome biology 23, no. 1 (2022): 88.

## Aim

Investigate the similarities and differences between the different protocols.

## Read data {.tabset}

### Janjic et al Prime-seq

2 million read downsampling object is selected as this is the most Janjic et al., is sequenced. 

```{r}
dat_prime <- here::here("data/GenomicsRnD/PrimeSeq/prime-seq.dgecounts.rds")
zUMI_prime <- readRDS(dat_prime)
prime_counts <- zUMI_prime$umicount$exon$all

# Rename columns
colnames(prime_counts) <- paste("Janjic Prime", colnames(prime_counts), sep="_")
```

### In house Prime-seq

```{r}
dat_mb2 <- here::here("data/GenomicsRnD/S000514/zUMI_outputs/expression/MB2_UHRR_Gencode_Hs.dgecounts.rds")
zUMI_mb2 <- readRDS(dat_mb2)
mb2_counts <- zUMI_mb2$umicount$exon$all

# Keep the UHRR added colums only
mb2_counts <- mb2_counts[,colSums2(mb2_counts) > 5000]

# Rename columns
colnames(mb2_counts) <- paste("Modified Prime", colnames(mb2_counts), sep="_")
```

### Novel protocol TIRE-Seq

TIRE-seq which uses the Qiagen turbocapture plate

```{r turbo}
dat_turbo <- "/stornext/Projects/GenomicsRnD/brown.d/S000514/zUMI_outputs/expression/TIRE_UHRR_Gencode_Hs.dgecounts.rds"
zUMI_turbo <- readRDS(dat_turbo)
turbo_counts <- zUMI_turbo$umicount$exon$all

# Keep the UHRR added colums only
turbo_counts <- turbo_counts[,colSums2(turbo_counts) > 5000]

colnames(turbo_counts) <- paste("TIRE", colnames(turbo_counts), sep="_")
```

# Correlate samples

Need to subset for genes that are in common with all protocols. Then I am able to combine them all together into a single count matrix.

```{r}
keep_genes <- intersect(
  row.names(prime_counts),
  row.names(mb2_counts)
)

keep_genes <- intersect(
  keep_genes,
  row.names(turbo_counts)
)

diff_prime_turbo <- setdiff(
  row.names(prime_counts),
  row.names(turbo_counts)
)

diff_mb2turbo <- setdiff(
  row.names(mb2_counts),
  row.names(turbo_counts)
)

count_combo <- cbind(
  prime_counts[keep_genes,],
  mb2_counts[keep_genes,],
  turbo_counts[keep_genes,]
)
```

Make an annotation bar for the heatmap.

```{r}
annotation_df <- as.data.frame(colnames(
  count_combo
))

annotation_df$Protocol <- str_split(string = annotation_df$`colnames(count_combo)`, pattern = "_", simplify = T)[,1]
annotation_df$Tag_count <- recode(annotation_df$Protocol,
                              "Janjic Prime" = "Three prime",
                              "Modified Prime" = "Three prime",
                              "TIRE" = "Five prime")

annotation_df <- annotation_df[,c(2:3)]
```

## Generate heatmap

Generate the correlation matrix then plot as a heatmap.

```{r heatmap}
cor_matrix <- cor(as.matrix(count_combo))
rownames(annotation_df) <- colnames(cor_matrix)
row.names(cor_matrix) <- str_split(colnames(cor_matrix), pattern = "_", simplify = T)[,1]

pheatmap(cor_matrix,
         color = viridis(100),
         annotation_col = annotation_df,
         show_rownames = TRUE,
         show_colnames = FALSE,
         cluster_rows = FALSE,
         fontsize_number = 14,
         fontsize_row = 10,)
```

### Session info
```{r}
sessionInfo()
```
