---
title: "cDC2 timecourse"
description: "cDC2 timecourse from total culture"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/cDC2_time/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())

suppressPackageStartupMessages({
  library(SingleCellExperiment)
  library(tidyverse)
  library(scran)
  library(scater)
  library(edgeR)
  library(splines)
  library(patchwork)
  library(knitr)
  library(ggvenn)
  library(AnnotationDbi)
  library(org.Mm.eg.db)
  library(here)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
})
```

# cDC2 timecourse from total culture

## Recap

Prior 96w evaluation of TurboCapture-Seq v2 showed low UMIs recovered and high seq saturation. I did low throughput troubleshooting and didnâ€™t see any issues. Process this experiment myself taking care to remove residual liquids from wash steps.

Process Hui Shi of Naik lab Bcor + Flt3 timecourse. Includes a few of my samples.

Samples

* Sorted dendritic cells
* HEK293T cell lysates in 1x TCL @ cells/uL
* PBMC cell lysates in 1x TCL @ 500 cells/uL
* No template control 1x TCL

## Notebook recap

SCE object in generate in 1A_generateSCE_reads notebook. The samples were then clustered in the 2B Clustering Wt notebook

## Notebook aim

Compare different subsets sorted from dendritic cells at 5 and 7 days. The easiest contrasts to interpret are the difference from the common dendritic cell precursor (CDP).

Workflow is from:  
https://bioconductor.org/packages/devel/workflows/vignettes/RNAseq123/inst/doc/limmaWorkflow.html#differential-expression-analysis

## Read SCE and preprocessing

This was generated in notebook 2B.

```{r}
sce <- readRDS(here::here(
   "data/TIRE_dendritic_mouse/SCEs", "DCs_cluster.sce.rds"))

dge <- scran::convertTo(sce, type="edgeR")
dge_orig <- dge
```

Have a look at the important metadata.  
There are not enough replicates for the preDC contrast to be meaningful.

```{r}
tb <- dge$samples[,c("Cell_type", "Cell_number", "Ligand", "Timepoint_Day")]

tb %>% 
  dplyr::count(Cell_type, Ligand, Timepoint_Day)
```

## Recap the PCA 

The cell type is the major difference between samples with the timepoint being much less so.  
Write the day as text.

```{r pca_day}
pca_tb <- as_tibble(reducedDim(sce))
pca_tb$Cell_Type <- sce$Cell_type
pca_tb$Timepoint <- sce$Timepoint_Day
pca_tb$Timepoint <- as.factor(pca_tb$Timepoint)

plt2 <- ggplot(pca_tb, aes(x=PC1, y=PC2, colour=Cell_Type, label=Timepoint)) +
  geom_text(size=5) +
  xlab("PC1 (34%)") + ylab("PC2 (28%)") +
  scale_colour_brewer(palette = "Dark2") +
  theme_Publication()
  
plt2
```

## Filter samples and low expressed genes

Interested in only the total DC cultures and cDC2s. Remove the other cell types.

```{r remove_samples}
keep_samples <- c("Total_DC_culture", "cDC2")
dge <- dge[,dge$samples$Cell_type %in% keep_samples]
```

Doing this reduces the multiple testing burden and fits variation better. After half the genes are removed leaving 10,000.

```{r remove_genes}
dim(dge)
keep.exprs <- filterByExpr(dge, group=dge$samples$Cell_type, min.count=1)
dge <- dge[keep.exprs,, keep.lib.sizes=FALSE]
dim(dge)
```

Correct for composition biases by computing normalization factors with the trimmed mean of M-values method.

```{r norm_factors}
dge <- calcNormFactors(dge)
```

# Look at timecourse of differntiation from total DC culture to cDC2

Looking at the PCA and deciding on a subset to focus on cDC2 seems like a good choice:

* Is a mature subset unlike CDP and therefore no further on trajectory to go
* Has some variation between day 5 and day 7

Day 5:  
Expansion of DC progenitors and early differentiation into immature DCs.
Presence of a heterogeneous mix of progenitor cells and early DC subsets.
Limited functional capacity.

Day 7:  
Peak numbers of mature DCs with distinct cDC1, cDC2, and pDC populations.
Enhanced antigen presentation and cytokine production abilities.
Optimal time point for harvesting DCs for functional assays.

Set up a spline based analysis

## Fit spline

Use a cubic regression spline curve with 3 degrees of freedom.

```{r bcv}
dge$samples$Timepoint_Day <- as.numeric(as.character(dge$samples$Timepoint_Day))
splines <- ns(dge$samples$Timepoint_Day, df = 3)

design <- model.matrix(~splines, dge$samples)
head(design)
```

The three coefficients do not have any particular meaning. Hypothesis testing would only make sense if the three coefficients are assessed together. The advantage of using a cubic spline curve is that it provides more stable fit at the end points compared to a polynomial.

The spline curve with 3 degrees of freedonm has 2 knots where cubic polynomials are splined together. In general, choosing a number of degrees of freedom to be in range of 3-5 is reasonable.

### Fit BCV

Circle back to see why BCV has no dots.

```{r qld}
par(mfrow=c(1,2))

dge <- estimateDisp(dge, design)
summary(dge$trended.dispersion)

plotBCV(dge)

fit <- glmQLFit(dge, design, robust=TRUE)
summary(fit$var.prior)
plotQLDisp(fit)
```

## Timecourse analysis {.tabset}

In a time course experiment, we are looking for genes that change expression level over time. Here, the design matrix uses 3 natural spline basis vectors to model smooth changes over time, without assuming any particular pattern to the trend.  
We test for a trend by conducting F-tests on 3 df for each gene:

```{r test}
fit <- glmQLFTest(fit, coef=2:4)
```

The topTags function lists the top set of genes with most significant time effects.

```{r}
results <- as_tibble(as.data.frame(
  topTags(fit, n=length(fit$genes$Symbol))
  ))
results
```

The total number of genes with significant (5% FDR) changes at different doses can be examined with decideTests.

There are 3200 differentially expressed genes that change over the 7 days.

```{r}
summary(decideTests(fit))
```

Note that all three spline coefficients should be tested together in this way. It is not meaningful to replace the F-tests with t-tests for the individual coefficients, and similarly the logFC columns of the top table do not have any interpretable meaning.  
The trends should instead be interpreted by way of trend plots, as we show now.  
Finally, we visualize the fitted spline curves for the top four genes. We start by computing the observed and fitted log-CPM values for each gene:

```{r}
top_genes <- head(results$Symbol,4)

logCPM.obs <- as.data.frame(cpm(dge, log=TRUE, prior.count=fit$prior.count))
logCPM.obs$ID <- row.names(logCPM.obs)
logCPM.fit <- as.data.frame(cpm(fit, log=TRUE))
logCPM.fit$ID <- row.names(logCPM.fit)

# Mung the cpm columns into a single tibble
lcm.obs <- pivot_longer(logCPM.obs[top_genes,], cols = -ID, names_to = "sample", values_to = "count") %>% 
  rename(obs_cpm = count)
lcm.fit <- pivot_longer(logCPM.fit[top_genes,], cols = -ID, names_to = "sample", values_to = "count") %>% 
  rename(fit_cpm = count)

lcm.obs$fit_cpm <- lcm.fit$fit_cpm

# Add back the sample metadata
lcm <- left_join(lcm.obs, results[,c("ID", "Symbol")])
lcm <- left_join(lcm, dge$samples[,c("Well_BC", "Timepoint_Day")], by=c("sample" = "Well_BC"))
```

### Log scale visualisation

The hits are all MHCII genes unsurprisingly.

The CD74 gene encodes a protein known as the HLA class II histocompatibility antigen gamma chain, also referred to as the invariant chain (Ii) or CD74. This protein is crucial for the proper functioning of the immune system, particularly in antigen presentation by Major Histocompatibility Complex (MHC) class II molecules.

Any gene that begins with H2- in mice is part of the H-2 complex, the murine equivalent of the human MHC (called HLA). These genes encode proteins involved in the immune system's ability to recognize self from non-self.

Roles of H2- Genes:

* MHC Class I Genes (e.g., H2-K, H2-D): Present endogenous peptides (from within the cell) to CD8+ T cells.
* MHC Class II Genes (e.g., H2-Aa, H2-Ab1): Present exogenous peptides (from outside the cell) to CD4+ T cells.

```{r time_log, fig.asp=1}
ggplot(lcm) +
  geom_point(aes(y = obs_cpm, x=Timepoint_Day), color = "black") +
  geom_smooth(aes(y = fit_cpm, x=Timepoint_Day), method = "loess", se = FALSE, color = "red") +
  xlab("Time (days)") + ylab("logCPM") +
  theme_Publication(base_size = 22) + theme(axis.text.x = element_text(angle = 0, vjust = 0.5, hjust = 1)) +
  facet_wrap(~ ID, scales = "free_y", labeller = labeller(ID = label_value)) +
  theme(strip.text = element_text(face = "bold.italic"))
```

# Conclusion

The results make sense to me. Will confirm with the domain experts.
 
### Session info
```{r}
sessionInfo()
```