---
title: "S000553 T cell human Generate SCE object"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/generateSCE/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())

library(tidyverse)
library(scuttle)
library(scran)
library(scater)
library(here)
library(patchwork)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

# TIRE-Seq human T cell

## Recap

I sequenced these samples with Prime-seq on extracted RNAs already. Results were not that great. Here reprocess the same samples with TIRE-seq for a head to head comparison.

## Samples

* Yasmin Nouri plates.
* All wells have 10,000 cells in 50ul total volume.
* Comprised of 25ul cells in media and 25ul 2x buffer TCL
* The cell concentration is therefore 200 cells/uL.
* So 20uL input will be 4000 cells.

## Lab notes

The processing went as planned. Full writeup available at [ELN link](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/NDkxLjQwMDAwMDAwMDAwMDAzfDY4MTk3LzM3OC9UcmVlTm9kZS8zMzc4NjI0NXwxMjQ3LjM5OTk5OTk5OTk5OTk=)

# Generate count matrices for downstream analysis

Read in the sample sheets and metadata

```{r}
gene_names <- read_tsv(here::here(
  "/vast/projects/G000448_Protein_display/BulkRNA_Manuscript/TIRE/Tcell/expression/TIRE_Tcell.gene_names.txt"))

meta_data <- read_csv(here::here(
  "metadata/S000553/YN_TIRE_metadata_v2.csv"))
meta_data$Replicate <- as.factor(as.character(meta_data$Replicate))

mapping <- read_tsv(here::here(
  "/vast/projects/G000448_Protein_display/BulkRNA_Manuscript/TIRE/Tcell/stats/TIRE_Tcell.readspercell.txt"))

zUMI <- here::here("/vast/projects/G000448_Protein_display/BulkRNA_Manuscript/TIRE/Tcell/expression/TIRE_Tcell.dgecounts.rds")
```

I generated this manually based on the 2D plate map I wrote in Excel and added the sample barcodes with Excel vlookup.

Need to populate the rest of the metadata fields by splitting the sample ID column. 

```{r}
# Add Species
meta_data$Species <-"Human"  

meta_data$Subset <- str_split(meta_data$Sample_ID, "_", simplify = TRUE)[,1]
meta_data$Receptor <- str_split(meta_data$Sample_ID, "_", simplify = TRUE)[,2]
meta_data$Donor <- str_split(meta_data$Sample_ID, "_", simplify = TRUE)[,3]
meta_data$Timepoint <- str_split(meta_data$Sample_ID, "_", simplify = TRUE)[,4]

meta_data$Donor <- gsub(pattern = "D", replacement = "Donor_", meta_data$Donor)
meta_data$Timepoint <- gsub(pattern = "T", replacement = "Day_", meta_data$Timepoint)
meta_data$Cell_line <- "CAR"

meta_data %>% 
  dplyr::count(Subset, Donor, Timepoint, Receptor) 
```

Read the zUMI output

```{r read_zUMI}
zUMI <- readRDS(zUMI)

ds_names <- names(zUMI[["umicount"]][["exon"]][["downsampling"]])

# Check the number of samples for each downsampling
for (n in ds_names) {
  print(n)
  print(dim(zUMI[["umicount"]][["exon"]][["downsampling"]][[n]]))
}

counts <- zUMI[["umicount"]][["exon"]][["all"]]
dim(counts)
```

Compute sequencing saturation. At nearly 70% saturation it is not worth resequencing this sample.

```{r}
stat_umi <-  read_tsv(here::here(
  "/vast/projects/G000448_Protein_display/BulkRNA_Manuscript/TIRE/Tcell/stats/TIRE_Tcell.UMIcounts.txt"))
stat_umi <- stat_umi[c(1,2)]

mapping <- mapping %>% 
  group_by(RG) %>% 
  pivot_wider(names_from = type, values_from = N) %>% 
  mutate(
    Reads = rowSums(across(where(is.numeric))),
    Mapped = 100 - (Unmapped / Reads * 100),
    Exon_Map = (Exon / Reads) * 100
    )

stats <- left_join(mapping, stat_umi,
                   by=c("RG"="SampleID"))
stats$saturation <- 1 - (stats$Count / stats$Reads)
summary(stats$saturation)
```

### Attach metadata

All samples are represented in the count matrix and metadata.  

```{r}
meta_data %>% 
  dplyr::select(Well_BC, Sample_ID)
```

Attach mapping rate metadata

```{r mapping_rate}
meta_data <- left_join(meta_data, mapping,
                by = c("Well_BC" = "RG"))
```

Subset the sample barcodes for those used in the experiment.

```{r}
keep_samples <- intersect(colnames(counts), meta_data$Well_BC)
discard_samples <- setdiff(colnames(counts), meta_data$Well_BC)

dropped <- meta_data[!meta_data$Well_BC %in% keep_samples,]
counts <- counts[,keep_samples]
meta_data <- meta_data[meta_data$Well_BC %in% keep_samples,]

meta_data <- meta_data %>% 
  arrange(Well_BC)
```

Check the discarded samples from the metadata. There are none

```{r}
dropped
```

### Convert to SCE

```{r}
sce <- SingleCellExperiment(assays=list(counts=counts),
       colData = meta_data)
```

Add gene symbol to human

```{r}
gene_names <- gene_names %>% 
  arrange(gene_id)

gene_id <- as_tibble(row.names(sce))
colnames(gene_id) <- "gene_id"

geneID <- left_join(gene_id, gene_names,
                     by=join_by(gene_id))

rowData(sce)$ID <- geneID$gene_id
rowData(sce)$Symbol <- geneID$gene_name
row.names(sce) <- geneID$gene_name
```

### Compute QC metrics
Add per cell QC adds the QC info to the colData of the sce.

```{r}
is.mito.alt <- grepl("^MT", row.names(sce))

sce <- addPerCellQC(sce, subsets=list(
  Mito=is.mito.alt))
sce <- addPerFeatureQC(sce)
```

### Save SCE with quality control

```{r}
saveRDS(sce, here::here(
  "data/TIRE_Tcell/SCEs", "tcell_basic.sce.rds"))

tb <- as_tibble(colData(sce))

write_csv(tb, here::here(
  "data/TIRE_Tcell/Matrices/", "tcell_qc.csv"))
```

Save as a dgelist object

```{r}
dge <- convertTo(sce, type="edgeR")

saveRDS(dge, here::here(
  "data/TIRE_Tcell/SCEs", "tcell_basic.dge.rds"))
```

### Session info
```{r}
sessionInfo()
```