---
title: "Basic DE analysis T cells over time"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/de_time/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())
suppressPackageStartupMessages({
  library(SingleCellExperiment)
  library(tidyverse)
  library(scater)
  library(edgeR)
  library(pheatmap)
  library(viridis)
  library(ggrepel)
  library(splines)
  library(org.Hs.eg.db)
  library(patchwork)
  library(knitr)
  library(here)
  source(here::here("scripts/R/theme_publication.R"))
  source(here::here("scripts/R/create_geneset_barchart.R"))
theme_set(theme_Publication())
})
```

Define heatmap function

```{r}
create_gene_heatmap <- function(de_results_df, dge_object, font_size = 12, num_genes = 20) {
  # Subset the genes
  top_genes <- de_results_df$Symbol[1:num_genes]
  subset_matrix <- dge_object$logCPM[top_genes, colnames(dge_object$counts)]
  
  # Create annotation dataframe
  sample_info <- dge_object$samples
  annotation_df <- data.frame(
    Donor = sample_info$Donor,
    Subset = sample_info$Subset,
    Timepoint = sample_info$Timepoint
  )
  rownames(annotation_df) <- colnames(subset_matrix)
  
  # Reorder gene expression matrix
  ordered_samples <- order(sample_info$Timepoint, sample_info$Donor)
  subset_matrix <- subset_matrix[, ordered_samples]
  sample_info <- sample_info[ordered_samples, ]
  
  # Create the heatmap
  heatmap <- pheatmap(subset_matrix,
           scale = "row",
           cluster_rows = FALSE,
           cluster_cols = FALSE,
           color = viridis(50),
           show_colnames = FALSE,
           annotation_col = annotation_df,
           fontsize = font_size,
           silent = FALSE)
}
```

# TIRE-seq T cell human basic DE time

The aim of this notebook is to identify genes that differ between CD4 and CD8 subsets over time. This will be a basic DE analysis informed by the time course centric analysis conducted in notebook 3A splines timecourse.

## Recap

I sequenced these samples with Prime-seq on extracted RNAs already. Results were not that great. Here reprocess the same samples with TIRE-seq for a head to head comparison.

## Samples

* Yasmin Nouri plates.
* All wells have 10,000 cells in 50ul total volume.
* Comprised of 25ul cells in media and 25ul 2x buffer TCL
* The cell concentration is therefore 200 cells/uL.
* So 20uL input will be 4000 cells.

## Lab notes

The processing went as planned. Full writeup available at [ELN link](https://au-mynotebook.labarchives.com/share/Daniel_RnD_Projects/NDkxLjQwMDAwMDAwMDAwMDAzfDY4MTk3LzM3OC9UcmVlTm9kZS8zMzc4NjI0NXwxMjQ3LjM5OTk5OTk5OTk5OTk=)

# Read data

This was generated in *2A_clustering*


```{r read_sce}
sce <- readRDS(here::here(
     "data/TIRE_Tcell/SCEs/tcell_cluster.sce.rds"
))
# Reorder factor levels
sce$Timepoint <- factor(sce$Timepoint, 
                            levels = c("Day_0", "Day_1", "Day_2",
                                       "Day_5", "Day_6", "Day_8",
                                       "Day_9", "Day_13", "Day_15"))

keep_samples <-  c("Day_0", "Day_2","Day_15")
dge <- scran::convertTo(sce[,sce$Timepoint %in% keep_samples], type="edgeR")
tb <- as_tibble(dge$samples)

dge$samples$Timepoint <- droplevels(dge$samples$Timepoint)
```

Check on the perturbations conducted in this experiment:

```{r}
tb %>% 
  dplyr::count(Subset, Donor, Timepoint) %>% 
  arrange(Timepoint) %>% 
  head(10)
```

Filter for genes that have at least 5 counts.  
Currently keep 8,249

```{r filter_gene}
keep <- filterByExpr(dge, group=dge$samples$Timepoint, min.count=5)
dge <- dge[keep,]
summary(keep)
```

Correct for composition biases by computing normalization factors with the trimmed mean of M-values method.

```{r norm_factors}
dge <- calcNormFactors(dge)
```

## Basic DE analysis

Combine time and subset into a new factor. For an interaction analysis it is easier to be more explicit than having fancy multiplication terms i.e timepoint * subset

The intercept term in a design matrix represents the baseline or reference level in your experimental design. Specifically:

* When an intercept is included, it typically corresponds to the mean expression level of the reference group or condition.
* Other coefficients in the model are then interpreted as differences from this baseline.
* Have "0" in the model matrix means the intercept regressed out
* When you include an intercept term in the model matrix, it adds a column of 1s to the design matrix. This column of 1s allows the regression line to intersect the y-axis at a non-zero point.
  + Removing the intercept term (by using model.matrix(~0 + .)) forces the regression line to go through the origin (y-axis at 0).
* Including an intercept makes good sense when the first group represents a reference or control group, as all comparison are made with respect to this condition.
* In the case of this experiment, removing the intercept means I can make a later contrast matrix that explicitly lists the comparisons of interest.

```{r}
sm <- model.matrix(~0+Timepoint + Donor, data=dge$samples)
# hypens not allowed
colnames(sm) <- make.names(colnames(sm), unique = FALSE, allow_ = TRUE)
```

# Differential expression analysis {.tabset}

Decide on the contrasts.

Based on the timecourse centric analysis in notebook 3A_time_spline focus on the 2 timepoints with the mpst dynamic changes

* Day 0 vs Day 2
* Day 2 vs Day 15

```{r}
contr.matrix <- makeContrasts(
   Day0_vs_Day2 = TimepointDay_2 - TimepointDay_0,
   Day2_vs_Day15 = TimepointDay_15 - TimepointDay_2,
   levels = colnames(sm))

contr.matrix %>% 
  kable()
```

In each contrast, the format is A - B where:

* A represents the condition considered as the "treatment" or point of interest
* B represents the condition considered as the "control" or baseline

So in this analysis the positive log fold changes will be the later of the 2 timepoints being compared.

Fit this model using limma/ voom.

```{r}
par(mfrow=c(1,2))
v <- voom(dge, sm, plot=TRUE)

vfit <- lmFit(v, sm)
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)
efit <- eBayes(vfit)
plotSA(efit, main="Final model: Mean-variance trend")
```

## Perform the differential expression analysis.

Check how many genes are differentially expressed

```{r}
summary(decideTests(efit))
```

# Visualise DE genes{.tabset}

## Day 2 vs Day 0

CD8 is highest which is a good sanity check.

```{r}
day2_vs_day0 <- topTable(efit, coef=1, n=length(efit$genes$ID), sort.by = "logFC")
results <- as_tibble(day2_vs_day0)
results$ID <- rownames(day2_vs_day0)

# add a column of NAs
results$DElabel <- FALSE
# if log2Foldchange > 1 or < -1 and pvalue < 0.05, set as "UP"
results$DElabel[results$logFC > 1 & results$adj.P.Val < 0.1] <- "Day2"
results$DElabel[results$logFC < -1 & results$adj.P.Val < 0.1] <- "Day0"
```

Add gene labels

```{r}
results$genelabels <- ""
results$genelabels[results$Symbol == "FGL2"] <- "FGL2"
results$genelabels[results$Symbol == "LINC00861"] <- "LINC00861"
results$genelabels[results$Symbol == "KLF2"] <- "KLF2"
results$genelabels[results$Symbol == "SDK2"] <- "SDK2"

results$genelabels[results$Symbol == "CDK1"] <- "CDK1"
results$genelabels[results$Symbol == "MKI67"] <- "MKI67"
results$genelabels[results$Symbol == "TOP2A"] <- "TOP2A"
results$genelabels[results$Symbol == "CDC45"] <- "CDC45"
```

All markers of proliferation in day 2

### Generate volcano plot

```{r day2_vs_day0_volcano}
plt1 <- ggplot(data=results, aes(x=logFC, y=-log10(adj.P.Val), colour=DElabel, label=genelabels)) + 
  geom_point(alpha=0.33, size=1.5) +
  geom_text_repel(size=4, colour="black") +
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1))) +
  scale_color_manual(values = c("darkblue", "darkorange", "grey"), name = "Upregulated") +
  geom_vline(xintercept = 1, linetype="dotted") + 
  geom_vline(xintercept = -1, linetype="dotted") +
  theme_Publication()

plt1
```

### Generate heatmap

```{r heatmap_d2_v_d0}
dge$logCPM <- cpm(dge, log=TRUE, prior.count=1)
dge_subset <- dge[,dge$samples$Timepoint %in% c("Day_0", "Day_2")]
create_gene_heatmap(de_results_df = results, dge_object = dge_subset, font_size = 12, num_genes=15)
```

### Gene set testing with camera

Need to convert geneIDs from ensembl to enterez

```{r}
geneids <- as.data.frame(v$genes$ID)
colnames(geneids) <- "ENSEMBL"

geneids$entrez <- mapIds(org.Hs.eg.db, keys = geneids$ENSEMBL, keytype = "ENSEMBL", column = "ENTREZID")

load("data/MSigDB/human_H_v5p2.rdata")
idx <- ids2indices(Hs.H,identifiers = geneids$entrez)
```

Geneset testing

```{r}
cam.day2_v_0 <- camera(v,idx,sm,contrast=contr.matrix[,1])
head(cam.day2_v_0,10)
```

Visualize as a barplot.  
Nothing that interesting proliferation a very strong signal

```{r geneset_barchart_day2_0}
geom_GeneSet_Barchart(cam.day2_v_0)
```

## Day 15 vs day 2

There are a set of highly significant genes switched off between day 15 and 2

```{r}
day15_v_2 <- topTable(efit, coef=2, n=length(efit$genes$ID), sort.by = "logFC")
results <- as_tibble(day15_v_2)
results$ID <- rownames(day15_v_2)

# add a column of NAs
results$DElabel <- FALSE
# if log2Foldchange > 1 or < -1 and pvalue < 0.05, set as "UP"
results$DElabel[results$logFC > 1 & results$adj.P.Val < 0.1] <- "Day15"
results$DElabel[results$logFC < -1 & results$adj.P.Val < 0.1] <- "Day2"
```

Add gene labels

```{r}
results$genelabels <- ""
results$genelabels[results$Symbol == "KRT1"] <- "KRT1"
results$genelabels[results$Symbol == "AL158071.3"] <- "AL158071.3"
results$genelabels[results$Symbol == "CCR2"] <- "CCR2"
results$genelabels[results$Symbol == "MMP25"] <- "MMP25"
results$genelabels[results$Symbol == "CD52"] <- "CD52"

results$genelabels[results$Symbol == "IRF8"] <- "IRF8"
results$genelabels[results$Symbol == "RGS16"] <- "RGS16"
results$genelabels[results$Symbol == "PMAIP1"] <- "PMAIP1"
results$genelabels[results$Symbol == "IFIT3"] <- "IFIT3"
results$genelabels[results$Symbol == "HSP90AA1"] <- "HSP90AA1"

```

### Generate volcano plot

```{r day15_v_2_volcano}
plt2 <- ggplot(data=results, aes(x=logFC, y=-log10(adj.P.Val), colour=DElabel, label=genelabels)) + 
  geom_point(alpha=0.33, size=1.5) +
  geom_text_repel(size=4, colour="black") +
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1))) +
  scale_color_manual(values = c("darkorange", "darkblue","grey"), name = "Upregulated") +
  geom_vline(xintercept = 1, linetype="dotted") + 
  geom_vline(xintercept = -1, linetype="dotted") +
  theme_Publication()

plt2
```

### Generate heatmap

```{r heatmap_d15_v_2}
dge$logCPM <- cpm(dge, log=TRUE, prior.count=1)
dge_subset <- dge[,dge$samples$Timepoint %in% c("Day_15", "Day_2")]
create_gene_heatmap(de_results_df = results, dge_object = dge_subset, font_size = 12, num_genes=15)
```

### Gene set testing with camera

Nothing that interesting proliferation a very strong signal

```{r}
cam.day15_v_2 <- camera(v,idx,sm,contrast=contr.matrix[,2])
head(cam.day15_v_2,10)
```

Visualize as a barplot

```{r geneset_barchart_day15_2}
geom_GeneSet_Barchart(cam.day15_v_2)
```

### Session info
```{r}
sessionInfo()
```