---
title: "Human neurosphere dose"
description: "Follow the edgeR manual on time course analysis with splines"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/DE_spline_dose/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())

library(SingleCellExperiment)
library(tidyverse)
library(scuttle)
library(biomaRt)
library(edgeR)
library(scater)
library(splines)
library(patchwork)
library(here)
source(here::here("scripts/R/theme_publication.R"))
theme_set(theme_Publication())
```

## Recap

These are the first experiments where I use version 2 chemistry. Read 2 is now the barcode and UMI. Read 1 is the cDNA. This was done to avoid reading through the TSO on Ilumina read 1

This sequencing run consists of 3 experiments:

1. Test the version 2 protocol with magnetic poly-T Dynabeads [https://au-mynotebook.labarchives.com/share/Piper_Research_Project/MjAuOHwxNzE3NjUvMTYvVHJlZU5vZGUvNDE3OTUxMTA4N3w1Mi44](ELN link])
2. Benchmark the version 2 protocol with a mixture of human and mouse cell lines Daniel Brown generated [https://au-mynotebook.labarchives.com/share/Piper_Research_Project/MjkuOTAwMDAwMDAwMDAwMDAyfDE3MTc2NS8yMy9UcmVlTm9kZS8yMzM4NDg4MTE1fDc1Ljg5OTk5OTk5OTk5OTk5](ELN link)
3. Test version 2 protocol with a biological appliation from Zac Moore of Brain Cancer lab [https://au-mynotebook.labarchives.com/share/Piper_Research_Project/MzEuMjAwMDAwMDAwMDAwMDAzfDE3MTc2NS8yNC9UcmVlTm9kZS8zMzczNDM0NjE1fDc5LjE5OTk5OTk5OTk5OTk5](ELN link)

## Notebook aim

Differential expression testing. Use a spline based analysis from the edgeR manual.  
In notebook 3B I verbatim reproduce analysis Zac Moore sent me.

# Read SCE and extract col data

This object was generated in 2A_Clustering.  
Subset only day 3 timepoint for simplicity and look at varying doses.

```{r}
sce <- readRDS(here::here(
  "data/TIRE_brain_human/SCEs/", "brainCancer_subset.sce.rds"
))

dge <- scran::convertTo(sce, type="edgeR")

day <- 3
dge <- dge[,dge$samples$Day_Exposure %in% day]

tb <- as_tibble(dge$samples)
```

# Differential expression testing

Have a look at the important metadata.  

```{r}
tb %>% 
  dplyr::count(Drug, Day_Exposure, Dose_M) %>% 
  arrange(Day_Exposure, Dose_M)
```

Remove genes that are lowly expressed. 12,000 genes are kept

```{r gene_filter}
keep <- filterByExpr(dge, group=dge$samples$Drug, min.count=1)
dge <- dge[keep,]
summary(keep)
```

Correct for composition biases by computing normalization factors with the trimmed mean of M-values method.

```{r norm_factors}
dge <- calcNormFactors(dge)
```

Check the MDS plot

```{r}
limma::plotMDS(dge, labels=dge$samples$Dose_M)
```

## Perform differential expression testing

Use a cubic regression spline on dose with 3 degrees of freedom.

```{r bcv}
splines <- ns(dge$samples$Dose_M, df = 3)

design <- model.matrix(~splines, dge$samples)
head(design)
```

The three coefficients do not have any particular meaning. Hypothesis testing would only make sense if the three coefficients are assessed together. The advantage of using a cubic spline curve is that it provides more stable fit at the end points compared to a polynomial.

The spline curve with 3 degrees of freedom has 2 knots where cubic polynomials are splined together. In general, choosing a number of degrees of freedom to be in range of 3-5 is reasonable.

### Fit BCV

Compared to the edgeR example on timecourse analysis the biological variation is pretty low.

```{r qld}
par(mfrow=c(1,2))

dge <- estimateDisp(dge, design)
summary(dge$trended.dispersion)

plotBCV(dge)

fit <- glmQLFit(dge, design, robust=TRUE)
summary(fit$var.prior)
plotQLDisp(fit)
```

## Dose response trend analysis {.tabset}

In a dose response experiment, we are looking for genes that change expression level across doses. Here, the design matrix uses 3 natural spline basis vectors to model smooth changes over concentration, without assuming any particular pattern to the trend.  
We test for a trend by conducting F-tests on 3 df for each gene:

```{r test}
fit <- glmQLFTest(fit, coef=2:4)
```

The topTags function lists the top set of genes with most significant dose effects.

```{r}
results <- as_tibble(as.data.frame(
  topTags(fit, n=length(fit$genes$Symbol))
  ))
results
```

The total number of genes with significant (5% FDR) changes at different doses can be examined with decideTests.

There are 681 differentially expressed genes across dose.

```{r}
summary(decideTests(fit))
```

Note that all three spline coefficients should be tested together in this way. It is not meaningful to replace the F-tests with t-tests for the individual coefficients, and similarly the logFC columns of the top table do not have any interpretable meaning.  
The trends should instead be interpreted by way of trend plots, as we show now.  
Finally, we visualize the fitted spline curves for the top four genes. We start by computing the observed and fitted log-CPM values for each gene:

```{r}
top_genes <- head(results$Symbol,6)

logCPM.obs <- as.data.frame(cpm(dge, log=TRUE, prior.count=fit$prior.count))
logCPM.obs$ID <- row.names(logCPM.obs)
logCPM.fit <- as.data.frame(cpm(fit, log=TRUE))
logCPM.fit$ID <- row.names(logCPM.fit)

# Mung the cpm columns into a single tibble
lcm.obs <- pivot_longer(logCPM.obs[top_genes,], cols = -ID, names_to = "sample", values_to = "count") %>% 
  rename(obs_cpm = count)
lcm.fit <- pivot_longer(logCPM.fit[top_genes,], cols = -ID, names_to = "sample", values_to = "count") %>% 
  rename(fit_cpm = count)

lcm.obs$fit_cpm <- lcm.fit$fit_cpm

# Add back the sample metadata
lcm <- left_join(lcm.obs, results[,c("ID", "Symbol")])
lcm <- left_join(lcm, dge$samples[,c("Well_BC", "Dose_M")], by=c("sample" = "Well_BC"))
```

### Log scale

Generate the plot.  
It makes more sense to plot x axis in log scale

```{r dose_response_log}
custom_labels <- c("Vehicle", "100 nM", "1 \u03bcM", "10 \u03bcM", "100 \u03bcM")

ggplot(lcm) +
  geom_point(aes(y = obs_cpm, x=Dose_M), color = "black") +
  geom_smooth(aes(y = fit_cpm, x=Dose_M), method = "loess", se = FALSE, color = "red") +
  xlab("TMZ (M)") + ylab("logCPM") +
  scale_x_continuous(trans = "log10") + annotation_logticks(sides="b") + 
  theme_Publication() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  facet_wrap(~ID, scales = "free_y")
```

### Linear scale

Generate the plot linear x axis. The low doses are too compressed which is why the loess curve looks strange.

```{r dose_response_lin}
custom_labels <- c("Vehicle", "100 nM", "1 \u03bcM", "10 \u03bcM", "100 \u03bcM")

ggplot(lcm) +
  geom_point(aes(y = obs_cpm, x=Dose_M), color = "black") +
  geom_smooth(aes(y = fit_cpm, x=Dose_M), method = "loess", se = FALSE, color = "red") +
  xlab("TMZ (M)") + ylab("logCPM") +
  theme_Publication() + theme(axis.text.x = element_text(angle = 90, vjust = 0.5, hjust = 1)) +
  facet_wrap(~ID, scales = "free_y")
```

## Gene descriptions

* MDM2 gene encodes a protein that negatively regulates the p53 tumor suppressor. MDM2 binds to p53, inhibiting its transcriptional activity and promoting its degradation via the ubiquitin-proteasome pathway. This regulation is crucial for controlling cell cycle progression and apoptosis.
* HMGCS1 gene encodes 3-hydroxy-3-methylglutaryl-CoA synthase 1, an enzyme crucial for the mevalonate pathway, which is key in cholesterol biosynthesis and the production of other isoprenoids. 
  + HMGCS1 catalyzes the first step in this pathway, combining acetyl-CoA and acetoacetyl-CoA to form HMG-CoA. This gene's expression is regulated by various factors, including sterol levels. 
  + Alterations in HMGCS1 can impact lipid metabolism and are associated with metabolic disorders and diseases such as cancer.
* DDIT3 gene encodes a transcription factor known as CHOP (C/EBP homologous protein) or GADD153. This protein is involved in cellular stress responses, particularly endoplasmic reticulum (ER) stress. 
  + DDIT3 promotes apoptosis by regulating genes associated with cell death and differentiation. 
  + Its overexpression is linked to increased apoptosis, contributing to diseases such as diabetes, neurodegenerative disorders, and cancer. 
  + DDIT3's role in stress responses makes it a potential target for therapeutic intervention in related diseases.
* TXNIP encodes thioredoxin interacting protein (TXNIP), which is a key regulator of cellular redox balance and glucose metabolism. 
  + TXNIP binds to and inhibits thioredoxin, a protein involved in reducing oxidative stress. By modulating oxidative stress and glucose uptake, TXNIP plays a crucial role in various physiological and pathological processes, including diabetes, cancer, and cardiovascular diseases. 
  + Elevated levels of TXNIP are often associated with increased oxidative stress and apoptosis.

### Session info
```{r}
sessionInfo()
```