---
title: "Human neurosphere basic differential expression test"
author: "Daniel Brown"
date: "`r Sys.Date()`"
output:
  html_document:
    toc: true
    toc_depth: 3
    toc_float:
      smooth_scroll: true
    theme: readable
    highlight: tango 
    df_print: paged
    code_folding: hide
editor_options: 
    chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      dev = "png",
                      dpi = 300,
                      fig.asp=0.6, 
                      fig.path='./figures/DE_time/',
                      warning=FALSE, message=FALSE)
knitr::opts_knit$set(root.dir = here::here())

library(SingleCellExperiment)
library(tidyverse)
library(scuttle)
library(biomaRt)
library(edgeR)
library(org.Hs.eg.db)
library(scater)
library(ggrepel)
library(pheatmap)
library(viridis)
library(patchwork)
library(here)
source(here::here("scripts/R/theme_publication.R"))
source(here::here("scripts/R/create_geneset_barchart.R"))
theme_set(theme_Publication())
```

Heatmap function

```{r}
create_gene_heatmap <- function(de_results_df, dge_object, font_size = 12, num_genes = 20) {
  # Subset the genes
  top_genes <- de_results_df$Symbol[1:num_genes]
  subset_matrix <- dge_object$logCPM[top_genes, colnames(dge_object$counts)]
  
  # Create annotation dataframe
  sample_info <- dge_object$samples
  annotation_df <- data.frame(
    Dose = sample_info$Dose
  )
  rownames(annotation_df) <- colnames(subset_matrix)
  
  # Reorder gene expression matrix
  ordered_samples <- order(sample_info$Dose)
  subset_matrix <- subset_matrix[, ordered_samples]
  sample_info <- sample_info[ordered_samples, ]
  
  # Create the heatmap
  heatmap <- pheatmap(subset_matrix,
           scale = "row",
           cluster_rows = FALSE,
           cluster_cols = FALSE,
           color = viridis(50),
           show_colnames = FALSE,
           annotation_col = annotation_df,
           fontsize = font_size,
           silent = FALSE)
}
```

## Recap

These are the first experiments where I use version 2 chemistry. Read 2 is now the barcode and UMI. Read 1 is the cDNA. This was done to avoid reading through the TSO on Ilumina read 1

This sequencing run consists of 3 experiments:

Test version 2 protocol with a biological application from Zac Moore of Brain Cancer lab [https://au-mynotebook.labarchives.com/share/Piper_Research_Project/MzEuMjAwMDAwMDAwMDAwMDAzfDE3MTc2NS8yNC9UcmVlTm9kZS8zMzczNDM0NjE1fDc5LjE5OTk5OTk5OTk5OTk5](ELN link)

## Notebook aim

Differential expression testing. Focus on the timepoints that show the most dynamic changes.  
Keep the 10uM dose becuase in notebook 3B this caused moderate effects.

# Read SCE and extract col data

This object was generated in 2A_Clustering.  
Subset only 10uM dose for simplicity.

```{r}
sce <- readRDS(here::here(
  "data/TIRE_brain_human/SCEs/", "brainCancer_subset.sce.rds"
))

dge <- scran::convertTo(sce, type="edgeR")

dose <- c(0, 1.00e-05)
dge <- dge[,dge$samples$Dose_M %in% dose]

tb <- as_tibble(dge$samples)
```

Convert the numerical dose column to a factor. Keep the 10uM dose becuase in notebook 3B this caused moderate effects

```{r}
tb$Dose <- as.factor(tb$Dose_M)
tb$Dose <- recode(tb$Dose,
                  "0" = "DMSO",
                  "1e-05" = "10uM")
dge$samples$Dose <- tb$Dose

dge <- dge[,dge$samples$Dose %in% c("DMSO", "10uM")]
dge$samples$Dose <- droplevels(dge$samples$Dose)
```

Create a new factor for time and dose.  
This is simpler than an interaction term in the design matrix.

```{r}
dge$samples$Dose_Time <- paste(dge$samples$Dose, dge$samples$Day_Exposure, sep="_d")
```

# Differential expression testing

Have a look at the important metadata.  

```{r}
tb %>% 
  dplyr::count(Drug, Day_Exposure, Dose) %>% 
  arrange(Day_Exposure, Dose)
```

Remove genes that are lowly expressed. 11,000 genes are kept

```{r gene_filter}
keep <- filterByExpr(dge, group=dge$samples$Drug, min.count=1)
dge <- dge[keep,]
summary(keep)
```

Correct for composition biases by computing normalization factors with the trimmed mean of M-values method.

```{r norm_factors}
dge <- calcNormFactors(dge)
```

## Perform differential expression testing

Model matrix generation.

```{r bcv}
design <- model.matrix(~0 + Dose_Time, dge$samples)
head(design)

dge <- estimateDisp(dge, design)
summary(dge$trended.dispersion)
```

Create the contrast matrix. Foucs on the largest changes.

```{r}
contr.matrix <- makeContrasts(
   day5_vs_day3 = Dose_Time10uM_d5 - Dose_Time10uM_d3,
   day7_vs_day3 = Dose_Time10uM_d7 - Dose_Time10uM_d3,
    day7_vs_day5 = Dose_Time10uM_d7 - Dose_Time10uM_d5,
   levels = colnames(design))
```

In each contrast, the format is A - B where:

* A represents the condition considered as the "treatment" or point of interest
* B represents the condition considered as the "control" or baseline

So the baseline is always the earlier timepoint

Fit BCV

```{r qld}
par(mfrow=c(1,2))
v <- voom(dge, design, plot=TRUE)

vfit <- lmFit(v, design)
vfit <- contrasts.fit(vfit, contrasts=contr.matrix)
efit <- eBayes(vfit)
plotSA(efit, main="Final model: Mean-variance trend")
```

## Differential expression analysis

Check how many genes are differentially expressed.  
There are the most at the ealrier timepoint so we focus on day 5 vs day 3

```{r}
summary(decideTests(efit))
```

# Visualise differentially expressed genes {.tabset}

## Day 5 vs day 3

```{r}
day5 <- topTable(efit, coef=1, n=length(efit$genes$ID), sort.by = "logFC")
results <- as_tibble(day5)
results$ID <- rownames(day5)

# add a column of NAs
results$DElabel <- FALSE
# if log2Foldchange > 1 or < -1 and pvalue < 0.05, set as "UP"
results$DElabel[results$logFC > 1 & results$adj.P.Val < 0.1] <- "Day5"
results$DElabel[results$logFC < -1 & results$adj.P.Val < 0.1] <- "Day3"
```

Add gene labels

```{r}
results$genelabels <- ""
results$genelabels[results$Symbol == "CDCA5"] <- "CDCA5"
results$genelabels[results$Symbol == "PBX1"] <- "PBX1"
results$genelabels[results$Symbol == "LDLR"] <- "LDLR"
results$genelabels[results$Symbol == "EXO1"] <- "EXO1"
results$genelabels[results$Symbol == "SPON1"] <- "SPON1"
results$genelabels[results$Symbol == "ABCA5"] <- "ABCA5"
results$genelabels[results$Symbol == "MRPS6"] <- "MRPS6"
results$genelabels[results$Symbol == "NEAT1"] <- "NEAT1"
```

### Volcano

```{r volcano_day3_day5}
plt1 <- ggplot(data=results, aes(x=logFC, y=-log10(adj.P.Val), colour=DElabel, label=genelabels)) + 
  geom_point(alpha=0.33, size=1.5) +
  geom_text_repel(size=4, colour="black") +
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1))) +
  scale_color_manual(values = c("darkblue", "darkorange", "grey"), name = "Upregulated") +
  geom_vline(xintercept = 1, linetype="dotted") + 
  geom_vline(xintercept = -1, linetype="dotted") +
  theme_Publication()

plt1
```
### Gene set testing with camera

Need to convert geneIDs from ensembl to enterez

```{r}
geneids <- as.data.frame(v$genes$ID)
colnames(geneids) <- "ENSEMBL"

geneids$entrez <- mapIds(org.Hs.eg.db, keys = geneids$ENSEMBL, keytype = "ENSEMBL", column = "ENTREZID")
```

The things that go down are the typical cell cycle arrest stuff. 
What goes up is more interesting in the ANGIOGENESIS and protein secretion.

```{r}
load("data/MSigDB/human_H_v5p2.rdata")
idx <- ids2indices(Hs.H,identifiers = geneids$entrez)
cam.day5_3<- camera(v,idx,design,contrast=contr.matrix[,1])
head(cam.day5_3,10)
```

Visualize the gene set testing.

```{r hallmark_sigs_day5, fig.cap="(top)HALLMARK_ANGIOGENESIS \n (bottom)HALLMARK_E2F_TARGETS"}
par(mfrow=c(1,1))

barcodeplot(efit$t[,1], index=idx$HALLMARK_ANGIOGENESIS,
            index2 = idx$HALLMARK_E2F_TARGETS)
```

Visualize as a barplot

```{r geneset_barchart_day5}
geom_GeneSet_Barchart(cam.day5_3)
```

## day 7 vs day 3

```{r}
day7 <- topTable(efit, coef=2, n=length(efit$genes$ID), sort.by = "logFC")
results <- as_tibble(day7)
results$ID <- rownames(day7)

# add a column of NAs
results$DElabel <- FALSE
# if log2Foldchange > 1 or < -1 and pvalue < 0.05, set as "UP"
results$DElabel[results$logFC > 1 & results$adj.P.Val < 0.1] <- "Day7"
results$DElabel[results$logFC < -1 & results$adj.P.Val < 0.1] <- "Day3"
```

Add gene labels

```{r}
results$genelabels <- ""
results$genelabels[results$Symbol == "PBX1"] <- "PBX1"
results$genelabels[results$Symbol == "RPH3A"] <- "RPH3A"
results$genelabels[results$Symbol == "KIAA0391"] <- "KIAA0391"
results$genelabels[results$Symbol == "CENPP"] <- "CENPP"
results$genelabels[results$Symbol == "CHTF18"] <- "CHTF18"
results$genelabels[results$Symbol == "FOSL2"] <- "FOSL2"
```

### Volcano

```{r volcano_day3_day7}
plt1 <- ggplot(data=results, aes(x=logFC, y=-log10(adj.P.Val), colour=DElabel, label=genelabels)) + 
  geom_point(alpha=0.33, size=1.5) +
  geom_text_repel(size=4, colour="black") +
  guides(colour = guide_legend(override.aes = list(size=3, alpha=1))) +
  scale_color_manual(values = c("darkblue", "grey", "darkorange"), name = "Upregulated") +
  geom_vline(xintercept = 1, linetype="dotted") + 
  geom_vline(xintercept = -1, linetype="dotted") +
  theme_Publication()

plt1
```
### Gene set testing with camera

The things that go down are the typical cell cycle arrest stuff. 
What goes up is more interesting in the ANGIOGENESIS and hedgehog signalling.

```{r}
cam.day7_3<- camera(v,idx,design,contrast=contr.matrix[,2])
head(cam.day7_3,10)
```

Visualize the gene set testing.

```{r hallmark_sigs_day7, fig.cap="(top)HALLMARK_HEDGEHOG_SIGNALING \n (bottom)HALLMARK_E2F_TARGETS"}
par(mfrow=c(1,1))

barcodeplot(efit$t[,1], index=idx$HALLMARK_HEDGEHOG_SIGNALING,
            index2 = idx$HALLMARK_E2F_TARGETS)
```

Visualize as a barplot

```{r geneset_barchart_day7}
geom_GeneSet_Barchart(cam.day7_3)
```


### Session info
```{r}
sessionInfo()
```